public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx

```diff
@@ -0,0 +1,526 @@
+        const labels: SelectableLabel[] = languageProvider
+          .getLabelKeys()
+          .slice(0, MAX_LABEL_COUNT)
+          .map(label => ({ name: label, selected: selectedLabels.includes(label), loading: false }));
+        this.setState({ labels }, () => {
+          this.state.labels.forEach(label => {
+            if (label.selected) {
+              this.fetchValues(label.name);
+            }
+          });
+        });
+      });
+    }
+  }
+
+  doFacetting = (lastFacetted?: string) => {
+    const selector = buildSelector(this.state.labels);
+    if (selector === EMPTY_SELECTOR) {
+      // Clear up facetting
+      const labels: SelectableLabel[] = this.state.labels.map(label => {
+        return { ...label, facets: 0, values: undefined, hidden: false };
+      });
+      this.setState({ labels }, () => {
+        // Get fresh set of values
+        this.state.labels.forEach(label => label.selected && this.fetchValues(label.name));
+      });
+    } else {
+      // Do facetting
+      this.fetchSeries(selector, lastFacetted);
+    }
+  };
+
+  async fetchValues(name: string) {
+    const { languageProvider } = this.props;
+    this.updateLabelState(name, { loading: true });
+    try {
+      const values: FacettableValue[] = (await languageProvider.getLabelValues(name))
+        .slice(0, MAX_VALUE_COUNT)
+        .map(value => ({ name: value }));
+      this.updateLabelState(name, { values, loading: false });
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  async fetchSeries(name: string, lastFacetted?: string) {
+    const { languageProvider } = this.props;
+    // this.updateLabelState(name, { loading: true });
+    try {
+      const possibleLabels = await languageProvider.fetchSeriesLabels(name);
+      if (Object.keys(possibleLabels).length === 0) {
+        // Sometimes the backend does not return a valid set
+        console.error('No results for label combination, but should not occur.');
+        // TODO mark label as broken instead of returning here
+        return;
+      }
+      const labels: SelectableLabel[] = facetLabels(this.state.labels, possibleLabels, lastFacetted);
+      this.setState({ labels });
+    } catch (error) {
+      console.error(error);
+    }
+  }
+
+  render() {
+    const { theme } = this.props;
+    const { labels, searchTerm } = this.state;
+    if (labels.length === 0) {
+      return <LoadingPlaceholder text="Loading labels..." />;
+    }
+    const styles = getStyles(theme);
+    let matcher: RegExp;
+    let selectedLabels = labels.filter(label => label.selected && label.values);
+    if (searchTerm) {
+      // TODO extract from render() and debounce
+      try {
+        matcher = new RegExp(searchTerm.split('').join('.*'), 'i');
+        selectedLabels = selectedLabels.map(label => ({
+          ...label,
+          values: label.values?.filter(value => value.selected || matcher.test(value.name)),
+        }));
+      } catch (error) {}
```

Comment:
What error are we catching here? it seems like only error here could be programatic one.

Reply:
+1